crates/wdl-engine/src/eval.rs

```diff
@@ -49,6 +55,167 @@ const MAX_STDERR_LINES: usize = 10;
+const CANCEL_REASON_USER: u8 = 2;
+
+/// State used in a [`CancellationContext`].
+#[derive(Default)]
+struct CancellationContextState {
+    /// Whether or not the underlying cancellation token is currently
+    /// cancellable.
+    ///
+    /// This is initially `false` for the "slow" failure mode where the first
+    /// cancellation should transition to "fast" failure mode and the second
+    /// call to `cancel` should cancel the underlying cancellation token.
+    ///
+    /// This is initially `true` for the "fast" failure mode where cancellation
+    /// should occur on the first call to `cancel`.
+    cancelable: AtomicBool,
+    /// The cancellation reason.
+    ///
+    /// See the above reason constants.
+    reason: AtomicU8,
+}
+
+/// Represents context for cancelling workflow or task evaluation.
+///
+/// Uses a default failure mode of [`Slow`](FailureMode::Slow).
+#[derive(Clone)]
+pub struct CancellationContext {
+    /// The state of the cancellation context.
+    state: Arc<CancellationContextState>,
+    /// Stores the underlying cancellation token.
+    token: CancellationToken,
+}
+
+impl CancellationContext {
+    /// Constructs a cancellation context for the given [`FailureMode`].
+    ///
+    /// If the provided `mode` is [`Slow`](FailureMode::Slow), the context will
+    /// be initialized such that the first call to [`cancel`][Self::cancel] will
+    /// transition the context to a [`Fast`](FailureMode::Fast) failure mode.
+    ///
+    /// If the provided `mode` is [`Fast`](FailureMode::Fast), the first call to
+    /// [`cancel`](Self::cancel) will cancel the underlying cancellation token.
+    pub fn new(mode: FailureMode) -> Self {
+        Self {
+            state: Arc::new(CancellationContextState {
+                cancelable: match mode {
+                    FailureMode::Slow => false,
+                    FailureMode::Fast => true,
+                }
+                .into(),
+                reason: CANCEL_REASON_NONE.into(),
+            }),
+            token: CancellationToken::new(),
+        }
+    }
+
+    /// Determines if a cancellation has occurred.
+    pub fn is_canceled(&self) -> bool {
+        self.state.reason.load(Ordering::SeqCst) != CANCEL_REASON_NONE
+    }
+
+    /// Performs a cancellation.
+    ///
+    /// If the initial failure mode was [`Slow`](FailureMode::Slow), this will
+    /// transition the cancellation context to a [`Fast`](FailureMode::Fast)
+    /// failure mode. A subsequent call to [`cancel`][Self::cancel] will cancel
+    /// the underlying cancellation token.
+    ///
+    /// If the initial failure mode was [`Fast`](FailureMode::Fast), this will
+    /// immediately cancel the underlying cancellation token.
+    ///
+    /// Returns `true` if the underlying cancellation token was canceled
+    /// (i.e. currently `Fast` mode) or `false` if it was not (i.e. previously
+    /// `Slow` mode).
+    pub fn cancel(&self) -> bool {
```

Comment:
nit: it might be worth a `#[must_use]` on this method since it doesn't always perform a true cancellation.